### PR TITLE
Adds feature for squashing commits and importing a project as a new commit.

### PR DESCRIPTION
**Pull request by kecso**
The new features that come with this enhancement:
- squashCommits a new function on the storage API, that allows users to squash multiple commits and replace them with a single one containing all the changes (see git analogy)
- updateProject is a new function on the server side, that is extending the import capabilities (you are able to create a new commit from a given commit or branch with the help of a webgmex package that will be added to your project as the new state)

!!! updateProject have not been made available on the UI yet and the squash is automatic without proper confirmation dialog !!!

### CODE REVIEW

File: src/bin/import.js [hunk 1]
```diff
@@ -118,6 +119,7 @@ main = function (argv) {
         .option('-b, --branch [branch]',
             'the branch that should be created with the imported data [by default it is the \'master\']')
         .option('-w --overwrite [boolean]', 'if a project exist it will be deleted and created again')
+        .option('-c --commit [string]', 'if given, the commit will be updated with the import')
```
Comment (pmeijer) [inline src/bin/import.js]: This is a little bit misleading...

File: src/client/js/Widgets/ProjectRepository/ProjectRepositoryWidget.js [hunk 2]
```diff
@@ -947,6 +968,14 @@ define(['js/logger',
                 '<i class="glyphicon glyphicon-remove"></i></button>'),
             self = this;
 
+        //if squash, then show confirm dialog and be done with it
+        if (isSquash) {
+            self.onSquashFromCommit({
+                branchName: this._start,
+                commitId: btn.data(COMMIT_IT)
```
Comment (pmeijer) [inline src/client/js/Widgets/ProjectRepository/ProjectRepositoryWidget.js]: COMMIT_IT? a typo?

File: src/client/js/Widgets/ProjectRepository/ProjectRepositoryWidgetControl.js [hunk 3]
```diff
@@ -85,6 +85,24 @@ define(['js/logger'], function (Logger) {
...
+        self._view.onSquashFromCommit = function (params) {
+            var projectName = self._client.getActiveProjectId();
+            self._client.squashBranchFromCommit(
+                params.branchName,
+                params.commitId,
+                function (err, result) {
+                    if (err) {
+                        self._logger.error(err);
+                    }
+                    console.log(result);
```
Comment (pmeijer) [inline src/client/js/Widgets/ProjectRepository/ProjectRepositoryWidgetControl.js]: console..

File: src/client/js/Widgets/ProjectRepository/ProjectRepositoryWidgetControl.js [hunk 4]
```diff
@@ -85,6 +85,24 @@ define(['js/logger'], function (Logger) {
...
+                    if (result.status === 'SYNCED') {
+                        self._view._dialog.setSelectorValue(params.branchName);
+                    } else {
+                        self._view._dialog.setSelectorValue();
```
Comment (pmeijer) [inline src/client/js/Widgets/ProjectRepository/ProjectRepositoryWidgetControl.js]: Maybe make a client notification here?

File: src/server/storage/safestorage.js [hunk 5]
```diff
@@ -787,6 +786,57 @@ SafeStorage.prototype.makeCommit = function (data, callback) {
 };
 
 /**
+ * Authorization: write access for data.projectId
```
Comment (pmeijer) [inline src/server/storage/safestorage.js]: Fill out parameters

File: src/common/storage/storageclasses/editorstorage.js [hunk 6]
```diff
@@ -464,6 +464,17 @@ define([
             }
         };
 
+        this.squashCommits = function (projectId, fromCommit, toCommitOrBranch, msg, callback) {
```
Comment (pmeijer) [inline src/common/storage/storageclasses/editorstorage.js]: Would be nice to have this on project/userproject too.
Additionally on the REST api.

File: src/server/storage/storage.js [hunk 7]
```diff
@@ -312,6 +313,129 @@ Storage.prototype.makeCommit = function (data, callback) {
...
+    //first we should check if the common ancestor is right
+    var self = this,
+        project,
+        deferred = Q.defer(),
+        fromCommit = data.fromCommit,
+        branchName = undefined,
+        rootHash,
+        toCommit,
+        msg,
+        filterBasedOnReach = function (allCommitItems) {
```
Comment (pmeijer) [inline src/server/storage/storage.js]: Consider moving these two to storageHelpers

File: src/server/storage/userproject.js [hunk 8]
```diff
@@ -236,6 +236,19 @@ function UserProject(dbProject, storage, mainLogger, gmeConfig) {
         return storage.getCommonAncestorCommit(data)
             .nodeify(callback);
     };
+
+    this.squashCommits = function (fromCommit, toCommitOrBranch, message, callback) {
```
Comment (pmeijer) [inline src/server/storage/userproject.js]: Found it here :)

File: src/server/storage/storage.js (hunk 7 above)
Comment (kecso) [inline src/server/storage/storage.js]: right now there are no other use for them and they seems pretty specific, but I am open for discussion

File: src/server/storage/safestorage.js [hunk 9]
```diff
@@ -787,6 +786,61 @@ SafeStorage.prototype.makeCommit = function (data, callback) {
...
+            entityType: self.authorizer.ENTITY_TYPES.PROJECT
+        },
+        rejected = false;
+
+    rejected = check(data !== null && typeof data === 'object', deferred, 'data is not an object.') ||
+
+        check(typeof data.projectId === 'string', deferred, 'data.projectId is not a string.') ||
+        check(REGEXP.PROJECT.test(data.projectId), deferred, 'data.projectId failed regexp: ' + data.projectId) ||
+
+        check(data.fromCommit !== null && typeof data.fromCommit === 'string', deferred,
```
Comment (pmeijer) [inline src/server/storage/safestorage.js]: no need to check null.

File: src/server/storage/safestorage.js [hunk 10]
```diff
@@ -787,6 +786,61 @@ SafeStorage.prototype.makeCommit = function (data, callback) {
...
+            'data.fromCommit not a string.') ||
+        check(REGEXP.HASH.test(data.fromCommit), deferred, 'data.fromCommit failed regexp: ' + data.fromCommit) ||
+        check(data.toCommitOrBranch !== null && typeof data.toCommitOrBranch === 'string', deferred,
```
Comment (pmeijer) [inline src/server/storage/safestorage.js]: same here..

File: src/server/worker/workerrequests.js [hunk 11]
```diff
@@ -1111,6 +1112,81 @@ function WorkerRequests(mainLogger, gmeConfig) {
...
+            .nodeify(callback);
+    }
+
+    /**
+     *
+     * @param webgmeToken
+     * @param parameters
+     * @param callback
+     */
+    function squashCommits(webgmeToken, parameters, callback) {
```
Comment (pmeijer) [inline src/server/worker/workerrequests.js]: Do we really need a worker-request for this?
All work is done by the server in any case..

**Pull request merged**